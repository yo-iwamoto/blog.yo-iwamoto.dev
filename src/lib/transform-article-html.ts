import { type DOMWindow, JSDOM } from "jsdom";

const jsdom = new JSDOM();
const domParser = new jsdom.window.DOMParser();

/**
 * Transforms the HTML of an article.
 * - Adds an ID to each heading for permalinks.
 */
export function transformArticleHtml(html: string) {
  const document = domParser.parseFromString(html, "text/html");

  assignIdToHeadings(document);

  return document.body.innerHTML;
}

function assignIdToHeadings(document: DOMWindow["document"]) {
  const headings = Array.from(document.querySelectorAll("h2, h3, h4, h5, h6"));

  for (const heading of headings) {
    const text = heading.textContent;
    if (text === null) continue;

    const id = text.replaceAll(/ /g, "-").toLowerCase();
    heading.id = id;

    const parent = document.createElement("a");
    parent.href = `#${id}`;
    parent.appendChild(heading.cloneNode(true));
    parent.classList.add("heading-link");

    heading.replaceWith(parent);
  }
}
