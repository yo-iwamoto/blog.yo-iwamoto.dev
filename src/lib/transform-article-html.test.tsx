import { render, screen } from "@testing-library/react";
import { transformArticleHtml } from "./transform-article-html";

const html = `<h2>Heading 2</h2>
<p>Paragraph 1</p>
<ul>
<li>Item 1</li>
<li>Item 2</li>
</ul>
<h3>Heading 3</h3>
<p>Paragraph 2</p>
<ol>
<li>Item 1</li>
<li>Item 2</li>
</ol>`;

describe("transformArticleHtml", () => {
  it("見出しがアンカーリンクで囲われていること", () => {
    const transformed = transformArticleHtml(html);
    render(
      // biome-ignore lint/security/noDangerouslySetInnerHtml: body は HTML 形式で取得されるため
      <div dangerouslySetInnerHTML={{ __html: transformed }} />,
    );
    expect(screen.getByRole("link", { name: "Heading 2" })).toHaveAttribute(
      "href",
      "#heading-2",
    );
    expect(screen.getByRole("link", { name: "Heading 3" })).toHaveAttribute(
      "href",
      "#heading-3",
    );
  });

  it("見出し要素が存在すること", () => {
    const transformed = transformArticleHtml(html);
    render(
      // biome-ignore lint/security/noDangerouslySetInnerHtml: body は HTML 形式で取得されるため
      <div dangerouslySetInnerHTML={{ __html: transformed }} />,
    );
    expect(screen.getByRole("heading", { name: "Heading 2" })).toBeDefined();
    expect(screen.getByRole("heading", { name: "Heading 3" })).toBeDefined();
  });

  it("transform 結果の snapshot が一致すること", () => {
    const transformed = transformArticleHtml(html);
    expect(transformed).toMatchSnapshot();
  });
});
