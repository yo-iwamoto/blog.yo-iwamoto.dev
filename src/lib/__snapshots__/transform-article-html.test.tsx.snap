// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformArticleHtml > transform 結果の snapshot が一致すること 1`] = `
"<a href="#heading-2" class="heading-link"><h2 id="heading-2">Heading 2</h2></a>
<p>Paragraph 1</p>
<ul>
<li>Item 1</li>
<li>Item 2</li>
</ul>
<a href="#heading-3" class="heading-link"><h3 id="heading-3">Heading 3</h3></a>
<p>Paragraph 2</p>
<ol>
<li>Item 1</li>
<li>Item 2</li>
</ol>"
`;
